part2/stages/Functions/install-main

```diff
@@ -591,7 +594,18 @@ create_efi_boot_entries()
 
 copy_to_esp()
 {
-    local ESP=$(sfdisk -l -q | grep "EFI System" | awk '{ print $1 }')
+    # TARGET_DISK not defined in upgrades; locate our
+    # installation disk manually.
+    if [ -z "${TARGET_DISK}" ]; then
+        local part="$(vgs xenclient --noheadings -o pv_name | sed 's/\s*//g' | sed 's|^/dev/||g')"
+
+        # get disk from partition (e.g. sda2 --> sda). using
+        # readlink instead of simply stripping off trailing
+        # numbers to cover other disk formats (e.g. nvme)
+        local TARGET_DISK="$(readlink /sys/class/block/$part | awk -F/ '{ print $(NF-1) }')"
+    fi
+
+    local ESP=$(sfdisk -l -q "/dev/${TARGET_DISK}" | grep "EFI System" | awk '{ print $1 }')
```

Comment:
You could elminate `grep` with `sfdisk -l -q "/dev/${TARGET_DISK}" | awk '/EFI System/{ print $1 }'`

Reply:
I'm going to leave the grep call in there for the sole reason of the operation being much clearer to someone who doesn't fully know the awk language (like myself).

Reply:
Sounds good.